safekeeper/src/send_interpreted_wal.rs

```diff
@@ -561,6 +561,20 @@ impl InterpretedWalReader {
                         // Update internal and external state, then reset the WAL stream
                         // if required.
                         let senders = self.shard_senders.entry(shard_id).or_default();
+
+                        // Clean up any shard senders that have dropped out before adding the new
+                        // one. This avoids a build up of dead senders.
+                        while let Some(sender) = senders.last() {
```

Comment:
May as well just check all the senders and remove them, e.g. via `senders.retain()`. This can terminate early if there's a live sender after a bunch of dead ones.